app/services/totp_setter.go

```diff
@@ -0,0 +1,46 @@
+package services
+
+import (
+	"github.com/keratin/authn-server/app"
+	"github.com/keratin/authn-server/app/data"
+	"github.com/keratin/authn-server/lib/compat"
+	"github.com/pkg/errors"
+	"github.com/pquerna/otp/totp"
+)
+
+//TOTPSetter persists the TOTP secret to the accountID if code is correct
+func TOTPSetter(accountStore data.AccountStore, totpCache data.TOTPCache, cfg *app.Config, accountID int, code string) error {
+	if code == "" { //Fail early if code is empty
+		return FieldErrors{{"totp", ErrInvalidOrExpired}}
+	}
+
+	account, err := AccountGetter(accountStore, accountID)
+	if err != nil {
+		return err
+	}
+
+	secret, err := totpCache.LoadTOTPSecret(account.ID)
+	if err != nil {
+		return err
+	}
```

Comment:
> if the secret has expired from cache, would we learn about it from this error check or the next?

Reply:
If the secret has expired I believe the read on the encrypted blob store returns an empty string. This error would catch any problems with the actual redis read which would be passed to the caller and trigger a panic in the handler. (Not a field error.) The next error would catch the empty string because the TOTP token wouldn't be valid. If the redis secret expired it would fail to validate returning a field error.